src/storage/json.rs

```diff
@@ -1,210 +1,87 @@
-                        None => return Err(CKR_DEVICE_ERROR)?,
-                    }
-                }
-            }
-            match uid {
-                Some(u) => cache.store(&u, obj)?,
-                None => return Err(CKR_DEVICE_ERROR)?,
-            }
-        }
-        Ok(())
-    }
-
-    pub fn from_cache(cache: &mut Box<dyn Storage>) -> JsonToken {
-        let objs = cache.search(&[]).unwrap();
-        let mut jt = JsonToken {
-            objects: Vec::with_capacity(objs.len()),
-        };
-        for o in objs {
-            if !o.is_token() {
-                continue;
-            }
-            jt.objects.push(JsonObject::from_object(&o));
-        }
-
-        jt
-    }
-
-    pub fn save(&self, filename: &str) -> Result<()> {
-        let jstr = match to_string_pretty(&self) {
-            Ok(j) => j,
-            Err(e) => return Err(Error::other_error(e)),
-        };
-        match std::fs::write(filename, jstr) {
-            Ok(_) => Ok(()),
-            Err(e) => Err(Error::other_error(e)),
-        }
-    }
+mod objects {
+    include!("json_objects.rs");
 }
+use objects::*;
 
 #[derive(Debug)]
 pub struct JsonStorage {
     filename: String,
-    cache: Box<dyn Storage>,
+    cache: Box<dyn StorageRaw>,
 }
 
-impl Storage for JsonStorage {
-    fn open(&mut self, filename: &String) -> Result<()> {
-        self.filename = filename.clone();
-        let token = JsonToken::load(&self.filename)?;
-        token.prime_cache(&mut self.cache)
+impl StorageRaw for JsonStorage {
+    fn is_initialized(&self) -> Result<()> {
+        self.cache.is_initialized()
     }
-    fn reinit(&mut self) -> Result<()> {
-        // TODO
+    fn db_reset(&mut self) -> Result<()> {
+        // TODO: reset not implemented yet
         Ok(())
     }
+    fn open(&mut self) -> Result<()> {
+        let token = JsonObjects::load(&self.filename)?;
+        token.prime_store(&mut self.cache)
+    }
     fn flush(&mut self) -> Result<()> {
-        let token = JsonToken::from_cache(&mut self.cache);
+        let token = JsonObjects::from_store(&mut self.cache);
         token.save(&self.filename)
     }
     fn fetch_by_uid(&self, uid: &String) -> Result<Object> {
         self.cache.fetch_by_uid(uid)
     }
-    fn store(&mut self, uid: &String, obj: Object) -> Result<()> {
-        self.cache.store(uid, obj)?;
-        self.flush()
-    }
     fn search(&self, template: &[CK_ATTRIBUTE]) -> Result<Vec<Object>> {
         self.cache.search(template)
     }
+    fn store_obj(&mut self, obj: Object) -> Result<()> {
+        self.cache.store_obj(obj)?;
+        self.flush()
+    }
     fn remove_by_uid(&mut self, uid: &String) -> Result<()> {
         self.cache.remove_by_uid(uid)?;
         self.flush()
     }
 }
 
-pub fn json() -> Box<dyn Storage> {
-    Box::new(JsonStorage {
-        filename: String::from(""),
-        cache: memory::memory(),
-    })
+#[derive(Debug)]
+pub struct JsonDBInfo {
```

Comment:
just a random though, I think there should be some warning that the json db is not encrypted, which might not be obvious on the first sight. Not sure whether this would be good place to have it in the source code comments or start writing some manual page or just a paragraph in readme describing the current format/configuration.
I will probably open a follow-up issue or MR with the draft.

Reply:
Actually the documentation is tracked as #105, but the storage differences and limitations should be mentioned somewhere somehow.

Reply:
Frankly, my intention is to eventually drop support for jsondb, the only reason I keep it is that the code is used to load some of the testing data, and allows for some testing of different storage layers.

But if you look at the latest patch it is optional and I have not changed CI to build with jsondb enabled ...